pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/custom/CustomDataQueryClusterIntegrationTest.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.pinot.integration.tests.ClusterIntegrationTestUtils;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.builder.ControllerRequestURLBuilder;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+import org.apache.pinot.util.TestUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.AfterSuite;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeSuite;
+
+
+public abstract class CustomDataQueryClusterIntegrationTest extends BaseClusterIntegrationTest {
+  protected static final Logger LOGGER = LoggerFactory.getLogger(CustomDataQueryClusterIntegrationTest.class);
+
+  @BeforeSuite
+  public void setUpSuite()
+      throws Exception {
+    LOGGER.info("Setting up integration test suite");
+    TestUtils.ensureDirectoriesExistAndEmpty(_tempDir, _segmentDir, _tarDir);
+
+    // Start the Pinot cluster
+    startZk();
+    startController();
+    startBroker();
+    startServer();
+    LOGGER.info("Finished setting up integration test suite");
+  }
+
+  @AfterSuite
+  public void tearDownSuite()
+      throws Exception {
+    LOGGER.info("Tearing down integration test suite");
+    // Shutdown the Pinot cluster
+    stopServer();
+    stopBroker();
+    stopController();
+    stopZk();
+    FileUtils.deleteDirectory(_tempDir);
+    LOGGER.info("Finished tearing down integration test suite");
+  }
+
+  @BeforeClass
+  public void setUp()
+      throws Exception {
+    LOGGER.info("Setting up integration test class: {}", getClass().getSimpleName());
+    if (_controllerRequestURLBuilder == null) {
+      _controllerRequestURLBuilder = ControllerRequestURLBuilder.baseUrl("http://localhost:" + DEFAULT_CONTROLLER_PORT);
+    }
+    TestUtils.ensureDirectoriesExistAndEmpty(_tempDir, _segmentDir, _tarDir);
+    // create & upload schema AND table config
+    Schema schema = getSchema();
+    addSchema(schema);
+    TableConfig tableConfig = getTableConfig();
+    addTableConfig(tableConfig);
+
+    // create & upload segments
+    File avroFile = createAvroFile();
+    ClusterIntegrationTestUtils.buildSegmentFromAvro(avroFile, tableConfig, schema, 0, _segmentDir, _tarDir);
+    uploadSegments(getTableName(), _tarDir);
+
+    waitForAllDocsLoaded(60_000);
+    LOGGER.info("Finished setting up integration test class: {}", getClass().getSimpleName());
+  }
+
+  @AfterClass
+  public void tearDown()
+      throws IOException {
+    LOGGER.info("Tearing down integration test class: {}", getClass().getSimpleName());
+    dropOfflineTable(getTableName());
+    FileUtils.deleteDirectory(_tempDir);
+    LOGGER.info("Finished tearing down integration test class: {}", getClass().getSimpleName());
+  }
+
+  public String getZkUrl() {
+    return ZkStarter.getDefaultZkStr();
+  }
```

Comment:
Do we need to override `getZkUrl()` and `getBrokerBaseApiUrl()`?

Reply:
yes, cause the zk/controller/broker/servers are created in the first test class, for other tests, those controller/broker/server are not created.